sympy/assumptions/refine.py

```diff
@@ -393,6 +391,54 @@ def refine_matrixelement(expr, assumptions):
             return expr
         return MatrixElement(matrix, j, i)
 
+def refine_Add(expr, assumptions):
+    """
+    Handler for advanced addition refinement.
+
+    """
+    finite_args = []
+    infty_type = None
```

Comment:
```suggestion
    infinity_type = None
    for subexpr in expr.args:
        if subexpr in (S.NaN, S.ComplexInfinity, S.Infinity, S.NegativeInfinity):
            if infinity_type is not None and infinity_type != subexpr:
                return S.NaN

            infinity_type = subexpr

    if infinity_type is not None:
        return infinity_type

    return expr
```

This should implement the same logic, but is much simpler.